### DISCUSSION TITLE
How should we capture best practices?

### DISCUSSION THREAD

**Discussion by sndrs** [guardian/csnx#429] [Team Posts]
Defining best practice captures good ideas and challenges them at the same time, and we should probably do it.

- should they be enforced? facilitated? or just recommended?
- how do we keep uses of best practise in sync with changes to them

**Comment by JamieB-gu** [comment]
What's the scope of this? For example:

- Build tools (webpack, microbundle)
- Libraries and frameworks (TypeScript, React, Preact)
- Style guides (ESLint, prettier, airbnb, custom gu style guide?)
- CSS (BEM, SASS, styled components)
- DevOps (Node servers, EC2, Lambda, containers, process managers like PM2)
- Project structure (directories, config files)
- JS features (ES6, async/await)

Are all of the above on the table? Are there some areas that people would like to leave up to individual projects?

> should they be enforced? facilitated? or just recommended?

Is the idea that if we rigorously enforce them then our projects become more uniform and it's easier to move between them?

In principle I like this idea, but in practice I hit some pain points with it when building [`liveblog-rendering`](https://github.com/guardian/liveblog-rendering) on top of `example-rendering`. I had to alter the setup to get it working with the specifics of that project, and some of the tools were unfamiliar to me and slowed development down a bit.

I think if we take this approach we have to a) come to the recommendations carefully, gathering lots of feedback, and b) start simple with very few tools and strictures, adding new ones only if there's a strong justification. If we just decide on a big chunk of tooling and requirements up front we may alienate projects and see limited adoption.

**Comment by SiAdcock** [comment]
As further reading, there are some existing [client side best practices](https://github.com/guardian/recommendations/blob/master/client-side.md) documented in the lesser known [guardian/recommendations](https://github.com/guardian/recommendations) repo. We should look to adopt and update these. I know @mchv is keen to preserve this repository as the central source of coding recommendations at the Guardian.
